morphir/runtime/src/org/finos/morphir/runtime/ToMDMConcept.scala

```diff
@@ -0,0 +1,165 @@
+    apply(distributions, bindings)
+
+  def as[B]: ToMDMConcept[B] = new ToMDMConcept[B] {
+    override def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither =
+      self.apply(distributions, bindings)
+  }
+}
+
+object ToMDMConcept {
+
+  sealed trait NoEquivalentMDM         extends Throwable
+  case class Unsupported(msg: String)  extends NoEquivalentMDM
+  case class MissingAlias(msg: String) extends NoEquivalentMDM
+  type ToMDMConceptEither = Either[NoEquivalentMDM, Concept]
+
+  def apply[A](implicit toMDMConcept: ToMDMConcept[A]): ToMDMConcept[A] = toMDMConcept
+  // def summon[A]: SummonPartiallyApplied[A]                              = new SummonPartiallyApplied[A]
+
+  def toConceptConverter[A](f: => Concept): ToMDMConcept[A] = new ToMDMConcept[A] {
+    def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither = Right(f)
+  }
+  def toFailedConverter[A](f: => NoEquivalentMDM): ToMDMConcept[A] = new ToMDMConcept[A] {
+    def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither = Left(f)
+  }
+
+  implicit def uTypeToConcept(tpe: UType): ToMDMConcept[UType] =
+    tpe match {
+      case _: Type.ExtensibleRecord[_] => toFailedConverter(Unsupported("Extensible record type has no MDM equivalent"))
+      case _: Type.Function[_]         => toFailedConverter(Unsupported("Functions do not exist within MDM"))
+      case _: Type.Unit[_]             => toConceptConverter(Concept.Unit)
+      case IntRef()                    => toConceptConverter(Concept.Int32)
+      case Int16Ref()                  => toConceptConverter(Concept.Int16)
+      case Int32Ref()                  => toConceptConverter(Concept.Int32)
+      case Int64Ref()                  => toConceptConverter(Concept.Int64)
+      case StringRef()                 => toConceptConverter(Concept.String)
+      case BoolRef()                   => toConceptConverter(Concept.Boolean)
+      case CharRef()                   => toConceptConverter(Concept.Char)
+      case FloatRef()                  => toConceptConverter(Concept.Float)
+      case DecimalRef()                => toConceptConverter(Concept.Decimal)
+      case LocalDateRef()              => toConceptConverter(Concept.LocalDate)
+      case LocalTimeRef()              => toConceptConverter(Concept.LocalTime)
+      case OrderRef()                  => toConceptConverter(Concept.Order)
+      case MonthRef()                  => toConceptConverter(Concept.Month)
+      case DayOfWeekRef()              => toConceptConverter(Concept.DayOfWeek)
+      case Type.Variable(_, name) =>
+        new ToMDMConcept[UType] {
+          def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither =
+            Right(bindings(name)) // TODO: Error this up nice
```

Comment:
TODO... this smells of something not completed. What is going on here?

Reply:
That's a map lookup - as is it will just throw an exception if something's missing, not return the appropriate Either. (And yeah, I wanted sign off on the high level approach before I filled in those level of details)